[otel-data] Hide 10m and 60m aggregated metrics generated for the APM UI

gregkalapos opened the pull request:
Solves https://github.com/elastic/elasticsearch/issues/113551

Aligning otel-data with apm-data to hide 10m and 60m aggregated metrics generated by collector components. 

Tested with the e2e setup as well:


<img width="1746" alt="Screenshot 2024-10-03 at 18 21 01" src="https://github.com/user-attachments/assets/f4b09037-6ed7-45f6-b8a1-e95e324b3032">

gregkalapos committed:
Hide aggregated metrics generated for the APM UI

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/index-templates/metrics-service_destination.10m@template.yaml:
```diff
@@ -0,0 +1,48 @@
+      data_stream.type:
+        type: constant_keyword
+        value: metrics
+      metricset:
```
Originally I wanted to define `constant_keyword` under `resource.attributes` for `metricset.interval` and under `attributes` for `metricset.name`, since that's where these are stored. But that'd again run into the usual issue of wiping out existing mapping for `resource.attributes`.

But as it seems, this works as well - this is also covered by tests.

Just calling out to make sure there is no unwanted implication I may missed.

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/index-templates/metrics-service_destination.10m@template.yaml:
```diff
@@ -0,0 +1,48 @@
+          interval:
+            type: constant_keyword
+            value: 10m
+          name:
+            type: constant_keyword
+            value: service_destination
```
`apm-data` only sets `interval`, but not name. I think it's worth setting both - if we decide not to do so, we could merge the 10m and 60m yaml files.

felixbarny replied:
Sounds intriguing to be able to reduce the number of index templates. However that would not allow us to have different `@custom` component templates, like `metrics-service_destination.10m.otel@custom`.

Having said that, maybe that level of granularity isn't really needed and what's more important is to have a custom template per interval, such as `metrics-10m.otel@custom`, so that users can set a retention per granularity.

Related to that, I'm wondering whether we should also create a template for the 1m intervals. That would enable us to add a custom component template to manage the retention for the 1m aggregated metrics, as well as having a constant_keyword field for the interval.

Also, I think the proliferation of index templates increases the need for a test that verifies the uniformity and consistency of the index templates. Manually reviewing that isn't really feasible anymore.

gregkalapos replied:
> Having said that, maybe that level of granularity isn't really needed and what's more important is to have a custom template per interval

Yeah, I agree this would be the right level of granularity. I added those `@custom`s 

> Related to that, I'm wondering whether we should also create a template for the 1m intervals.

Yes! Added that as well.

gregkalapos committed:
Update 30_aggregated_metrics_tests.yml

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

gregkalapos commented on x-pack/plugin/otel-data/src/main/resources/index-templates/metrics-service_destination.10m@template.yaml:
```diff
@@ -0,0 +1,48 @@
+          name:
+            type: constant_keyword
+            value: service_destination
+    dynamic_templates:
+      - ecs_ip:
+          mapping:
+            type: ip
+          path_match: [ "ip", "*.ip", "*_ip" ]
+          match_mapping_type: string
+      - all_strings_to_keywords:
+          mapping:
+            ignore_above: 1024
+            type: keyword
+          match_mapping_type: string
```
Should we do something with these? We repeat them everywhere.

If I remember correctly, the idea was to implement these here in the template, so people can't overwrite these.

First idea would be to move them to `metrics-otel@mappings.yaml`, but that won't cover what we want.

felixbarny replied:
Yes, I think it would be better to have a separate component template for these. 

> If I remember correctly, the idea was to implement these here in the template, so people can't overwrite these.

I think it was the opposite. Users should have the ability to customize the dynamic mappings. Therefore, when extracting this to a component template, we need to import that template after the @custom component template.

gregkalapos replied:
Yes, so what I mean: users should be able to change dynamic templates currently defined in `metrics-otel@mappings.yaml`, but not the ones above (`ecs_ip` and `all_strings_to_keywords`), because changing or removing those could easily break the data stream since it effects dimensions. Right?

So I'd move these two dynamic templates into a separate  components (maybe `metrics-otel-fixed@mappings`)  and metrics template would be something like:

```
composed_of:
  - metrics@tsdb-settings
  - otel@mappings
  - metrics-otel@mappings
  - semconv-resource-to-ecs@mappings
  - metrics@custom
  - metrics-otel@custom
  - metrics-otel-fixed@mappings #<- the new one
 ```
 
 (plus adding `metrics-[x]m.otel@custom` according to https://github.com/elastic/elasticsearch/pull/114042#discussion_r1787408686)

Does that sound ok?  

* Happy to take suggestion on the naming `metrics-otel-fixed@mappings` :)

gregkalapos commented:
Strange failing tests... so far what I see:

```
distribution/bwc/minor/build/bwc/checkout-8.x/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/OpenAiService.java:318: error: cannot find symbol
 [8.16.0] > Task :x-pack:plugin:vector-tile:compileJava
 [8.16.0] > Task :x-pack:plugin:vector-tile:classes
 [8.16.0] > Task :x-pack:plugin:vector-tile:jar
 [8.16.0]     public Set<TaskType> supportedStreamingTasks() {
 [8.16.0]            ^
 [8.16.0]   symbol:   class Set
 [8.16.0]   location: class OpenAiService
 [8.16.0] 
 [8.16.0] > Task :x-pack:plugin:inference:compileJava
```

Doesn't seem to be related to this PR.

gregkalapos commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into otel_hide_10_and_60m_aggregated_metrics

felixbarny commented on x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/30_aggregated_metrics_tests.yml:
```diff
@@ -0,0 +1,193 @@
+      indices.get_data_stream:
+        name: metrics-transaction.60m.otel-default
+  - match: { data_streams.0.hidden: true }
+
```
An additional generic test we could do here is to do a terms aggregation on metricset.interval. We'd expect only one bucket with the value "1m". When setting `expand_wildcards:open,hidden`, the aggregation is supposed to return all three values.

gregkalapos replied:
👍 Added these tests for all 4 aggregation indices.

gregkalapos committed:
Review feedback

- Introduced templates for 1 minute aggregations
- Moved dynamic templates `ecs_ip` and `all_strings_to_keywords` into a dedicated file and now I pull the file in instead of repeating them
- Introduced `metrics-[x]m.otel@custom`
- Added tests with terms aggregation that assert by default 1 bucket (only 1m) with metricsset.interval, and with allowing hidden indices it's 3 buckets (1m, 10m, 60m)

elasticmachine committed:
Merge branch 'main' into otel_hide_10_and_60m_aggregated_metrics

gregkalapos committed:
Update 30_aggregated_metrics_tests.yml

Simplify - no need to separate tests for the 3bucket queries.

felixbarny commented on x-pack/plugin/otel-data/src/main/resources/component-templates/metrics-otel-fixed@mappings.yaml:
```diff
@@ -0,0 +1,19 @@
+version: ${xpack.oteldata.template.version}
+_meta:
+  description: |
+    Default mappings that can't be changed by users for
+    the OpenTelemetry metrics index template installed by x-pack
```
It's the reverse. Users should be able to customize the dynamic mappings. That's why this component template needs to be imported last. This is because for dynamic templates, the first one that matches wins. Therefore, users need to be able to put a dynamic template before this in order to be able to customize how attributes are mapped dynamically. We have a test for this scenario:

https://github.com/elastic/elasticsearch/blob/3b9150dbb064a4afe3eb3602e9478d15eb72e409/x-pack/plugin/otel-data/src/yamlRestTest/resources/rest-api-spec/test/20_metrics_tests.yml#L112-L148

So I think all is correctly implemented, but we'll need  to change the description, and probably the name of the dynamic template. Suggestion: `ecs-tsdb@mappings` (since the ip path match is coming from ECS).

gregkalapos committed:
Rename metrics-otel-fixed@mappings to ecs-tsdb@mappings

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/114173)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/114173">](https://github.com/elastic/elasticsearch/pull/114173) |
